packages/devtools_app/lib/src/profiler/cpu_profile_model.dart

```diff
@@ -16,82 +16,166 @@ import 'cpu_profile_transformer.dart';
-        time: (json[timeOriginKey] != null && json[timeExtentKey] != null)
-            ? (TimeRange()
-              ..start = Duration(microseconds: json[timeOriginKey])
-              ..end = Duration(
-                  microseconds: json[timeOriginKey] + json[timeExtentKey]))
-            : null,
-      ),
+      stackFrames: stackFrames,
+      cpuSamples: samples,
+      profileMetaData: profileMetaData,
     );
   }
 
   factory CpuProfileData.subProfile(
     CpuProfileData superProfile,
     TimeRange subTimeRange,
   ) {
-    // Each trace event in [subTraceEvents] will have the leaf stack frame id
-    // for a cpu sample within [subTimeRange].
-    final subTraceEvents = superProfile.stackTraceEvents
-        .where((trace) => subTimeRange
-            .contains(Duration(microseconds: trace[TraceEvent.timestampKey])))
+    // Each sample in [subSamples] will have the leaf stack
+    // frame id for a cpu sample within [subTimeRange].
+    final subSamples = superProfile.cpuSamples
+        .where((sample) => subTimeRange
+            .contains(Duration(microseconds: sample.timestampMicros)))
         .toList();
 
     // Use a SplayTreeMap so that map iteration will be in sorted key order.
-    final SplayTreeMap<String, Map<String, dynamic>> subStackFramesJson =
+    final SplayTreeMap<String, CpuStackFrame> subStackFrames =
         SplayTreeMap(stackFrameIdCompare);
-    for (Map<String, dynamic> traceEvent in subTraceEvents) {
-      // Add leaf frame.
-      final String leafId = traceEvent[stackFrameIdKey];
-      final Map<String, dynamic> leafFrameJson =
-          superProfile.stackFramesJson[leafId];
-      subStackFramesJson[leafId] = leafFrameJson;
+    for (final sample in subSamples) {
+      final leafFrame = superProfile.stackFrames[sample.leafId];
+      subStackFrames[sample.leafId] = leafFrame;
 
       // Add leaf frame's ancestors.
-      String parentId = leafFrameJson[parentIdKey];
+      String parentId = leafFrame.parentId;
       while (parentId != null) {
-        final parentFrameJson = superProfile.stackFramesJson[parentId];
-        subStackFramesJson[parentId] = parentFrameJson;
-        parentId = parentFrameJson[parentIdKey];
+        final parentFrame = superProfile.stackFrames[parentId];
+        subStackFrames[parentId] = parentFrame;
+        parentId = parentFrame.parentId;
       }
     }
 
     return CpuProfileData._(
-      stackFramesJson: subStackFramesJson,
-      stackTraceEvents: subTraceEvents,
+      stackFrames: subStackFrames,
+      cpuSamples: subSamples,
       profileMetaData: CpuProfileMetaData(
-        sampleCount: subTraceEvents.length,
+        sampleCount: subSamples.length,
         samplePeriod: superProfile.profileMetaData.samplePeriod,
         stackDepth: superProfile.profileMetaData.stackDepth,
         time: subTimeRange,
       ),
     );
   }
 
+  factory CpuProfileData.fromUserTag(CpuProfileData originalData, String tag) {
```

Comment:
the code for this factory constructor was copied from what used to be the `dataForUserTag` method below. There are also some changes for using the new format of CpuProfileData